lib/channel-postmessage/src/index.ts

```diff
@@ -117,6 +164,39 @@ export class PostmsgTransport {
   }
 }
 
+const getEventSourceUrl = (event: MessageEvent) => {
+  const frames: HTMLIFrameElement[] = [...document.getElementsByTagName('iframe')];
+  let result = event.origin;
+  let frame;
+  let remainder;
+  try {
+    // try to find the originating iframe by matching it's contentWindow
+    // This might not be cross-origin safe
+    [frame, ...remainder] = frames.filter(element => element.contentWindow === event.source);
+
+    const src = frame.getAttribute('src');
+    const { pathname, origin } = new URL(src);
+
+    result = origin + pathname;
+  } catch (e) {
+    // if the contentWindow access was not allowed we match on origin alone
+    [frame, ...remainder] = frames.filter(element => {
+      const src = element.getAttribute('src');
+      const { origin } = new URL(src);
+
+      return origin === event.origin;
+    });
+  }
```

Comment:
It seems a bit weird that we will immediately fallback to the second method if *any* of the iframes don't allow CORS access. What about something more like the following:

```js
const [frame, ...remainder] = frames.filter(element => {
  try {
    return element.contentWindow === event.source;
  } catch (err) {
    // typical case, the iframe has not redirected at all
    return element.getAttribute('src') === event.source.document.location.href
  }
});

if (remainder.length > 0) throw new Error("something's weird");

if (frame) return [frame]

// We couldn't find the frame matching on window or current URL, see if there is a single frame with matching origin
const [frame, ...remainder] = frames.filter(e => new URL(element.getAttribute('src')).origin === event.origin);
if (frame && remainder.length === 0) return frame;

return null;
```

WDYT?

Reply:
I tweaked to WDYT of it now?

Reply:
It’s better but I still think we could check the `document.location.href` inside the iframe. Would that not work?

Reply:
that's not valid according to the types of the EventSource, which is why I didn't do it..

Reply:
I don't really get that? [MDN says](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) that `event.source` is:

> A reference to the window object that sent the message; you can use this to establish two-way communication between two windows with different origins.

So can't we use that to get a reference to the location? Haven't tried it obviously.